fix: not able do see reports about deleted users

ggazzo opened the pull request:
<!-- This is a pull request template, you do not need to uncomment or remove the comments, they won't show up in the PR text. -->

<!-- Your Pull Request name should start with one of the following tags
  feat: Adding a new feature
  refactor: A code change that doesn't change behavior (it doesn't add anything and doesn't fix anything)
  fix: For bug fixes that affect the end-user
  chore: For small tasks
  docs: For documentation
  ci: For updating CI configuration
  test: For adding tests
  i18n: For updating any translations
  regression: Issues created/reported/fixed during the development phase. kind of problem that never existed in production and that we don't need to list in a changelog for the end user
-->

<!-- Checklist!!! If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code. 
  - I have read the Contributing Guide - https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat doc
  - I have signed the CLA - https://cla-assistant.io/RocketChat/Rocket.Chat
  - Lint and unit tests pass locally with my changes
  - I have added tests that prove my fix is effective or that my feature works (if applicable)
  - I have added necessary documentation (if applicable)
  - Any dependent changes have been merged and published in downstream modules
-->

## Proposed changes (including videos or screenshots)
<!-- CHANGELOG -->
<!--
  Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
  If it fixes a bug or resolves a feature request, be sure to link to that issue below.
  This description will appear in the release notes if we accept the contribution.
-->

<!-- END CHANGELOG -->

## Issue(s)
<!-- Link the issues being closed by or related to this PR. For example, you can use #594 if this PR closes issue number 594 -->

## Steps to test or reproduce
<!-- Mention how you would reproduce the bug if not mentioned on the issue page already. Also mention which screens are going to have the changes if applicable -->

## Further comments
<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->

Predecessor PR: https://github.com/RocketChat/Rocket.Chat/pull/29701 
[WM-120](https://rocketchat.atlassian.net/browse/WM-120)


[WM-120]: https://rocketchat.atlassian.net/browse/WM-120?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Dnouv commented:
Made some changes for following:

1. User actions on the Moderation Console table e.g. "Deactivate user" was not disabled.
C. Added back the hook `useIsUserDeleted` which is being called inside the `ModerationConsoleAction`

2. The GUI crashes if reports are being checked (inside a context bar) of a user, who's never been reported.
C. Some additional optional chaining in the hook, which is getting the username, or name

3. Type errors on the `cloneAndSaveAsHistoryByRecord`
C. added a `// @ts-expect-error`

Thank you!

Dnouv commented:
I had to force push because I did a `develop` merge and reverted the same commit. But the commit history was still showing up, so removed all of it using a rebase.
Thank you!

KevLehman commented on apps/meteor/app/api/server/v1/moderation.ts:
```diff
@@ -73,10 +73,7 @@ API.v1.addRoute(
-			if (!user) {
-				return API.v1.failure('error-invalid-user');
-			}
+			const user = await Users.findOneById(userId, { projection: { _id: 1, username: 1, name: 1 } });
```
```suggestion
			const user = await Users.findOneById<Pick<IUser, 'username' | 'name' | '_id'>>(userId, { projection: { _id: 1, username: 1, name: 1 } });
```

Dnouv replied:
Thank you!

KevLehman commented on apps/meteor/server/lib/moderation/deleteReportedMessages.ts:
```diff
@@ -22,7 +22,14 @@ export async function deleteReportedMessages(messages: IMessage[], user: IUser):
+			for await (const doc of cursor) {
+				await Messages.cloneAndSaveAsHistoryByRecord(
+					doc,
+					user as Required<Pick<IUser, '_id' | 'username' | 'name'>> & { username: NonNullable<IUser['username']> },
```
```suggestion
					user as Required<Pick<IUser, '_id' | 'name'>> & { username: NonNullable<IUser['username']> },
```

Dnouv replied:
Thanks a lot for the review!

ggazzo commented:
thanks for the additions  @Dnouv,  but I deleted the hook for a good reason (actually 3)

- during the load state, the callout is going to be shown, blinking/disappearing after the user data arrives which is a bad ux
- one fetch per row potentially will  trigger the rate limiter if you have many reports
- if the user doesnt exist or if you get a rate limiter error react query will retry at least 3x, in a huge list you are going to fetch a bunch of times
I added this info to the aggregation we are going to be forced to check if the user exists anyway, so I believe it doesn’t add extra load to the operation

Dnouv commented:
I agree; we can remove them. Thank you so much for the changes! I missed this case. Indeed I noticed the query resulted in an error for the deleted users, and it was being called thrice in a row. 

Can the delay between two consecutive retries be modified (just asking out of curiosity)?

ggazzo commented:
> I agree; we can remove them. Thank you so much for the changes! I missed this case. Indeed I noticed the query resulted in an error for the deleted users, and it was being called thrice in a row.
> 
> Can the delay between two consecutive retries be modified (just asking out of curiosity)?
yes, react query allow us to change the behaviour

KevLehman commented:
@Dnouv , maybe you can update the api tests for this case? 🙏🏽  (on a separate PR so we don't have this staying open for longer)

Dnouv commented:
> yes, react query allow us to change the behaviour

Interesting, thank you so much!

> @Dnouv , maybe you can update the api tests for this case? 🙏🏽 (on a separate PR so we don't have this staying open for longer)

Sure, I would be happy to do that. Just confirming, do we need this test PR today?

ggazzo commented:
@Dnouv 
> Sure, I would be happy to do that. Just confirming, do we need this new test PR today?

no need
